pl_bolts/datamodules/experience_source.py

```diff
@@ -29,174 +31,247 @@ def __iter__(self) -> Iterable:
+
+        return actions
+
+    def env_step(self, env_idx: int, env: Env, action: List[int]) -> Experience:
         """
-        get the accumulated transition info for the n_step_buffer
+        Carries out a step through the given environment using the given action
+
         Args:
-            gamma: discount factor
+            env_idx: index of the current environment
+            env: env at index env_idx
+            action: action for this environment step
 
         Returns:
-            multi step reward, final observation and done
+            Experience tuple
         """
-        last_experience = self.n_step_buffer[-1]
-        final_state = last_experience.new_state
-        done = last_experience.done
-        reward = last_experience.reward
+        next_state, r, is_done, _ = env.step(action[0])
+
+        self.cur_rewards[env_idx] += r
+        self.cur_steps[env_idx] += 1
 
-        # calculate reward
-        # in reverse order, go through all the experiences up till the first experience
-        for experience in reversed(list(self.n_step_buffer)[:-1]):
-            reward_t = experience.reward
-            new_state_t = experience.new_state
-            done_t = experience.done
+        exp = Experience(state=self.states[env_idx], action=action[0], reward=r, done=is_done, new_state=next_state)
 
-            reward = reward_t + self.gamma * reward * (1 - done_t)
-            final_state, done = (new_state_t, done_t) if done_t else (final_state, done)
+        return exp
 
-        return reward, final_state, done
+    def update_env_stats(self, env: gym.Env, env_idx: int) -> None:
+        """
+        To be called at the end of the history tail generation during the termination state. Updates the stats
+        tracked for all environments
 
+        Args:
+            env: current environment to be reset
+            env_idx: index of the environment used to update stats
+        """
+        self.total_rewards.append(self.cur_rewards[env_idx])
+        self.total_steps.append(self.cur_steps[env_idx])
+        self.cur_rewards[env_idx] = 0
+        self.cur_steps[env_idx] = 0
+        self.states[env_idx] = env.reset()
 
-class EpisodicExperienceStream(ExperienceSource, IterableDataset):
-    """
-    Basic experience stream that iteratively yield the current experience of the agent in the env
+    def pop_total_rewards(self) -> List[float]:
+        """
+        Returns the list of the current total rewards collected
 
-    Args:
-        env: Environmen that is being used
-        agent: Agent being used to make decisions
-    """
+        Returns:
+            list of total rewards for all completed episodes for each environment since last pop
+        """
+        rewards = self.total_rewards
 
-    def __init__(self, env, agent, device: torch.device, episodes: int = 1):
-        super().__init__(env, agent)
-        self.episodes = episodes
-        self.device = device
+        if rewards:
```

Comment:
The total_rewards is a list that contains the latest rewards at the end of an episode. This gets called at the end of each step in our model. If the model calls pop_rewards_step, if an episode has finished, total_rewards will contain a value or if multiple episodes have finished since we last check, there will be multiple values. These get store in reward, then  we clear the total_rewards. We only want to empty self.total_rewards once we have taken the latest values. Does that make sense?